aten/src/ATen/core/jit_type.h

```diff
@@ -1068,4 +1079,111 @@ matchTypeVariables(TypePtr formal, TypePtr actual, TypeEnv& type_env);
 
 CAFFE2_API TypePtr evalTypeVariables(TypePtr type, TypeEnv & type_env);
 
+/**
+ * User Defined Types
+ */
+
+struct UserType;
+using UserTypePtr = std::shared_ptr<UserType>;
+using ::torch::jit::script::Module;
+using ::torch::jit::script::Method;
+
+// This represents a user-defined type in TorchScript.
+struct CAFFE2_API UserType : public Type {
+  // Create a user type and register it globally.
+  static UserTypePtr create(const std::string& name, std::shared_ptr<Module> module);
+  // returns nullptr if there is no type with that name
+  static UserTypePtr get(const std::string& name);
+
+  DEFINE_IS_SUBCLASS(UserType);
+  bool operator==(const Type& rhs) const override {
+    if (auto user_rhs = rhs.cast<UserType>()) {
+      return typename_ == user_rhs->typename_;
+    }
+    return false;
+  }
+
+  bool isSubtypeOf(const TypePtr rhs) const override {
+    // XXX: We do not have inheritance implemented, only types that are the
+    // same can subtype from each other.
+    return *this == *rhs;
+  }
+  std::string str() const override {
+    return std::string("UserType<") + typename_ + ">";
+  }
+
+  TypePtr getAttribute(const std::string& name) const {
+    const auto it = std::find_if(
+        attributes_.cbegin(), attributes_.cend(), [&](const Attribute& attr) {
+          return attr.name == name;
+        });
+    if (it == attributes_.cend()) {
+      return nullptr;
+    }
+
+    return it->type;
+  }
+
+  Method* getMethod(const std::string& name) const;
+
+  std::string name() const {
+    return typename_;
+  }
+
+  size_t numAttributes() const {
+    return attributes_.size();
+  }
+
+  // Attributes are stored in a specific slot at runtime for effiency.
+  // When emitting instructions we specify the slot so that attribute access is
+  // a constant lookup
+  size_t getAttributeSlot(const std::string& name) const {
```

Comment:
I feel like both the type names and attribute names could get interned safely. Do we really feel like people will be creating lots of those dynamically?

Reply:
When we are emitting getattr and setattr instructions we're getting the names in string form from the AST anyway, so the interface needs to accept strings at least. I think it's okay not to intern since it's only one string per attribute per type.